File: DataInjectorAndroid/DataInjectorLib/src/main/java/com/crescentflare/datainjector/transformer/UrlEncodeTransformer.java
package com.crescentflare.datainjector.transformer;

import com.crescentflare.datainjector.conversion.InjectorConv;
import com.crescentflare.datainjector.injector.DataInjector;
import com.crescentflare.datainjector.utility.InjectorPath;
import com.crescentflare.datainjector.utility.InjectorResult;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

/**
 * Data injector transformation: URL encode a string
 * Convert data to a URL encoded string
 */
public class UrlEncodeTransformer extends BaseTransformer
{
    // --
    // Members
    // --

    private InjectorPath sourceDataPath;


    // --
    // Initialization
    // --

    public UrlEncodeTransformer()
    {
    }


    // --
    // Manual transformation
    // --

    @NotNull
    public static InjectorResult encode(@Nullable Object data)
    {
        String string = InjectorConv.asString(data);
        if (string != null)
        {
            try
            {
                String result = URLEncoder.encode(string, "UTF8");
                return InjectorResult.withModifiedObject(result);
            }
            catch (UnsupportedEncodingException ignored)
            {
            }
        }
        return InjectorResult.withError(InjectorResult.Error.SourceInvalid);
    }


    // --
    // General transformation
    // --

    @Override
    @NotNull
    protected InjectorResult onApply(@Nullable Object sourceData)
    {
        Object useSourceData = DataInjector.get(sourceData, sourceDataPath != null ? sourceDataPath : new InjectorPath());
        return encode(useSourceData);
    }


    // --
    // Set values
    // --

    public void setSourceDataPath(@Nullable InjectorPath sourceDataPath)
    {
        this.sourceDataPath = sourceDataPath;
    }
}

File: DataInjectorIOS/DataInjector/Classes/transformer/UrlEncodeTransformer.swift
//
//  UrlEncodeTransformer.swift
//  DataInjector Pod
//
//  Library transformer: URL encode a string
//  Convert data to a URL encoded string
//

import Foundation

/// A transformer to URL encode a string
open class UrlEncodeTransformer: BaseTransformer {
    
    // --
    // MARK: Members
    // --
    
    public var sourceDataPath: InjectorPath?

    
    // --
    // MARK: Initialization
    // --
    
    public override init() {
    }

    
    // --
    // MARK: Manual transformation
    // --
    
    public static func encode(fromData: Any?) -> InjectorResult {
        if let dataString = InjectorConv.asString(value: fromData) {
            let characters = (CharacterSet.urlQueryAllowed as NSCharacterSet).mutableCopy() as! NSMutableCharacterSet
            characters.removeCharacters(in: "!*'();:@&=+$,/?%#[]")
            characters.addCharacters(in: " ")
            guard let encodedString = dataString.addingPercentEncoding(withAllowedCharacters: characters as CharacterSet) else {
                return InjectorResult(withModifiedObject: dataString)
            }
            return InjectorResult(withModifiedObject: encodedString.replacingOccurrences(of: " ", with: "+"))
        }
        return InjectorResult(withError: .sourceInvalid)
    }


    // --
    // MARK: General injection
    // --
    
    open override func appliedTransformation(sourceData: Any?) -> InjectorResult {
        let useSourceData = DataInjector.get(from: sourceData, path: sourceDataPath ?? InjectorPath(path: ""))
        return UrlEncodeTransformer.encode(fromData: useSourceData)
    }

}
